addons/payment/controllers/portal.py

```diff
@@ -100,7 +100,8 @@ def payment_pay(
         reference = reference or payment_utils.singularize_reference_prefix(prefix='tx')
         amount = amount or 0.0  # If the amount is invalid, set it to 0 to stop the payment flow
         company_id = company_id or partner_sudo.company_id.id or user_sudo.company_id.id
-        currency_id = currency_id or request.env['res.company'].browse(company_id).currency_id.id
+        res_company = request.env['res.company'].sudo().browse(company_id)
```

Comment:
just nitpicking here, but  when `company_id` is not set (is it even possible?) you are browsing the same record as above, so this could be written as
``` python
res_company = company_id and request.env['res.company'].sudo().browse(company_id) or partner_sudo.company_id or user_sudo.company_id
company_id = res_company.id
currency_id = currency_id or res_company.currency_id.id
```